File: src/main/java/io/gittul/domain/bookmark/entity/BookmarkService.java
package io.gittul.domain.bookmark.entity;

import io.gittul.domain.post.PostRepository;
import io.gittul.domain.post.entity.Post;
import io.gittul.domain.user.UserRepository;
import io.gittul.domain.user.entity.User;
import io.gittul.global.exception.CustomException;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class BookmarkService {

    private final PostRepository postRepository;
    private final UserRepository userRepository;

    public void addBookmark(User user, Long postId) {
        Post post = postRepository.getReferenceById(postId);
        Bookmark bookmark = Bookmark.of(user, post);

        if (post.isBookmarkedBy(user)) throw new CustomException(HttpStatus.CONFLICT, "이미 북마크한 게시글입니다.");

        user.getDetails().getBookmarks().add(bookmark);
        userRepository.save(user);
    }

    public void removeBookmark(User user, Long postId) {
        user.getDetails()
                .getBookmarks()
                .removeIf(bookmark -> bookmark.getPost().getPostId().equals(postId));

        userRepository.save(user);
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/bookmark/BookmarkService.kt
package io.gittul.app.domain.bookmark

import io.gittul.app.domain.thread.repository.ThreadRepository
import io.gittul.app.domain.user.UserRepository
import io.gittul.core.domain.bookmark.entity.Bookmark
import io.gittul.core.domain.user.entity.User
import io.gittul.core.global.exception.CustomException
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class BookmarkService(
    private val threadRepository: ThreadRepository,
    private val userRepository: UserRepository
) {

    fun addBookmark(user: User, threadId: Long) {
        val thread = threadRepository.getReferenceById(threadId)
        val bookmark = Bookmark.of(user, thread)

        if (thread.isBookmarkedBy(user)) throw CustomException(HttpStatus.CONFLICT, "이미 북마크한 게시글입니다.")

        user.details.bookmarks.add(bookmark)
        userRepository.save(user)
    }

    fun removeBookmark(user: User, threadId: Long) {
        user.details
            .bookmarks
            .removeIf { it.thread.threadId == threadId }

        userRepository.save(user)
    }
}
